### PR TITLE
Check disabled connections after protocol update

### LINKED ISSUE #16043: Re-enable connections invalidated due to protocol version
## What

When we upgrade/downgrade a connector, we should re-enable connections that were invalided due to protocol version if the new protocol version is back in the supported range.

## How

This should happen around the API where we get the spec of a new connector

### PR DESCRIPTION
**Pull request by gosusnp**
## What
Following a connector update, we want to re-enable connections that may have been disabled due to an unsupported protocol version if possible.

Closes #16043 

## How
Upon successful a connector update, clear unsupported protocol version flags if the protocol version is back within support range.

## Recommended reading order
1. `StandardSyncPersistence.java` contains the actual logic to re-enable connection
2. Rest would be up to the reviewer.

## 🚨 User Impact 🚨
No breaking change, we can start to re-enable connection that would have been disabled due to protocol version, however, the disable step doesn't exist yet so no impact here.

### CODE REVIEW

**Comment by alovew** [comment]
So for connections that have unsupported_protocol_version=true, we are not updating the status to inactive=true? we just rely on that field to determine whether it is disabled or not?

**Comment by gosusnp** [comment]
> So for connections that have unsupported_protocol_version=true, we are not updating the status to inactive=true? we just rely on that field to determine whether it is disabled or not?

@alovew, yes, the desired behavior is that after a connector update, if it fixes some protocol version conflicts, we'd automatically restore the connection to the state they were in before the conflict. We opted for adding a flag to mark the issue rather than changing the actual status. It means that we need to update how we lookup connections to take this into consideration. This will be part of the work to actual disable connections which is going to be my next task.

File: airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StandardSyncPersistence.java [hunk 1]
```diff
@@ -214,4 +253,40 @@ private void writeStandardSync(final StandardSync standardSync, final DSLContext
...
+    final ActorDefinition sourceDef = ACTOR_DEFINITION.as("sourceDef");
+    final ActorDefinition destDef = ACTOR_DEFINITION.as("destDef");
+
+    // Retrieve all the connections currently disabled due to a bad protocol version
+    // where the actor definition is matching the one provided to this function
+    final Stream<StandardSyncIdsWithProtocolVersions> results = ctx
+        .select(CONNECTION.ID, sourceDef.ID, sourceDef.PROTOCOL_VERSION, destDef.ID, destDef.PROTOCOL_VERSION)
+        .from(CONNECTION)
+        .join(source).on(CONNECTION.SOURCE_ID.eq(source.ID))
+        .join(sourceDef).on(source.ACTOR_DEFINITION_ID.eq(sourceDef.ID))
```
Comment (benmoriceau) [inline airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StandardSyncPersistence.java:269]: Why do we need the join on the source / source def ? I think that it is missing in the where clause.

File: airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StandardSyncPersistence.java [hunk 2]
```diff
@@ -214,4 +253,40 @@ private void writeStandardSync(final StandardSync standardSync, final DSLContext
...
+        .join(destination).on(CONNECTION.DESTINATION_ID.eq(destination.ID))
+        .join(destDef).on(destination.ACTOR_DEFINITION_ID.eq(destDef.ID))
+        .where(
+            CONNECTION.UNSUPPORTED_PROTOCOL_VERSION.eq(true).and(
+                (actorType == ActorType.DESTINATION ? destDef : sourceDef).ID.eq(actorDefId)))
```
Comment (benmoriceau) [inline airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StandardSyncPersistence.java:274]: Should we do the same for the Source Actor type?

File: airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StandardSyncPersistence.java [hunk 3]
```diff
@@ -214,4 +253,40 @@ private void writeStandardSync(final StandardSync standardSync, final DSLContext
...
+            r.get(CONNECTION.ID),
+            r.get(sourceDef.ID),
+            AirbyteProtocolVersion.getWithDefault(r.get(sourceDef.PROTOCOL_VERSION)),
+            r.get(destDef.ID),
+            AirbyteProtocolVersion.getWithDefault(r.get(destDef.PROTOCOL_VERSION))));
+    return results;
+  }
+
+  private void clearProtocolVersionFlag(final DSLContext ctx, final List<UUID> standardSyncIds) {
+    ctx.update(CONNECTION)
```
Comment (benmoriceau) [inline airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StandardSyncPersistence.java:286]: Nit: Please update the modified_at collumn as well. I don't think that is being updated through a trigger.

Comment (gosusnp) [reply airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StandardSyncPersistence.java:286]: Good catch, updated.

**Review by alovew — APPROVED** [review]

**Review by benmoriceau — APPROVED** [review]

**Pull request merged**